### ISSUE TITLE
Add ability to pause/resume execution.

### ISSUE THREAD

**Issue by PiotrSikora** [WebAssembly/wasm-c-api#76]
This might be V8 specific, but the current API seems to be designed around short-lived and non-overlapping Stores, where `v8::Isolate` is `Enter()'ed` upon creation, and `Exit()'ed` during destruction.

For long-lived Stores (e.g. WASM extensions in HTTP proxies), which can be created and destroyed at arbitrary times, this means jumping between Stores without explicit context switching (which might be safe?) and `Enter()` / `Exit()` calls in no particular order, e.g.:
```
auto store1 = wasm::Store::make(engine.get());  // store1->isolate()->Enter();
auto store2 = wasm::Store::make(engine.get());  // store2->isolate()->Enter();
store1.reset();                                 // store1->isolate()->Exit();
```
which trigger this `DCHECK` when V8 is built with `is_debug=true`:
```
DCHECK(CurrentPerIsolateThreadData()->isolate_ == this);
```

@jakobkummerow suggested `wasm::Store::pause()` and `wasm::Store::resume()` as possible APIs.

If we don't want to expose this at the API level, then we could do explicit `Enter()` / `Exit()` on each function call inside the library, but that sounds expensive.

Any thoughts, @rossberg?

**Comment by rossberg** [comment]
Ah, indeed, the implementation doesn't correctly support using multiple stores. The expectation is that you can, so I think we'll have to perform Enter/Exit in all the right places to address this. The problem is too V8-specific to justify something like pause/resume, which wouldn't serve any semantically apparent purpose.

@jakobkummerow, what's the estimated cost of Enter/Exit?

**Comment by jakobkummerow** [comment]
Accessing TLS and an allocation... it's not exactly cheap.

But we can handle this internally if needed. Short term, I'm hoping that we can find a way to call Enter/Exit only as needed; long term, I believe there's the option of cleaning up V8's API to the point where all functions that need it take an explicit Isolate* parameter, and Enter/Exit will become unnecessary. No idea about timeframe though, and feasibility becomes a certain fact only when someone actually implements it.

**Comment by PiotrSikora** [comment]
@jakobkummerow should we close this issue then? Any ETA for "costly" mitigation (i.e. Enter/Exit on each function call) in V8?

**Comment by jakobkummerow** [comment]
It's been addressed in V8 7.7. We can close this.
